contracts/ChainlinkPriceFeedV3.sol

```diff
@@ -0,0 +1,239 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.7.6;
+pragma abicoder v2;
+
+import { Address } from "@openzeppelin/contracts/utils/Address.sol";
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import { AggregatorV3Interface } from "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";
+import { IChainlinkPriceFeed } from "./interface/IChainlinkPriceFeed.sol";
+import { IChainlinkPriceFeedV3 } from "./interface/IChainlinkPriceFeedV3.sol";
+import { BlockContext } from "./base/BlockContext.sol";
+import { CachedTwap } from "./twap/CachedTwap.sol";
+
+contract ChainlinkPriceFeedV3 is IChainlinkPriceFeedV3, BlockContext, CachedTwap {
+    using SafeMath for uint256;
+    using Address for address;
+
+    //
+    // STATE
+    //
+
+    uint24 private constant _ONE_HUNDRED_PERCENT_RATIO = 1e6;
+    uint8 internal immutable _decimals;
+    uint24 internal immutable _maxOutlierDeviationRatio;
+    uint256 internal immutable _outlierCoolDownPeriod;
+    uint256 internal immutable _timeout;
+    uint256 internal _lastValidPrice;
+    uint256 internal _lastValidTimestamp;
+    AggregatorV3Interface internal immutable _aggregator;
+
+    //
+    // EXTERNAL NON-VIEW
+    //
+
+    constructor(
+        AggregatorV3Interface aggregator,
+        uint256 timeout,
+        uint24 maxOutlierDeviationRatio,
+        uint256 outlierCoolDownPeriod,
+        uint80 twapInterval
+    ) CachedTwap(twapInterval) {
+        // CPF_ANC: Aggregator is not contract
+        require(address(aggregator).isContract(), "CPF_ANC");
+        _aggregator = aggregator;
+
+        // CPF_IMODR: Invalid maxOutlierDeviationRatio
+        require(maxOutlierDeviationRatio < _ONE_HUNDRED_PERCENT_RATIO, "CPF_IMODR");
+        _maxOutlierDeviationRatio = maxOutlierDeviationRatio;
+
+        _outlierCoolDownPeriod = outlierCoolDownPeriod;
+        _timeout = timeout;
+        _decimals = aggregator.decimals();
+    }
+
+    /// @notice anyone can help with updating
+    /// @dev keep this function for PriceFeedUpdater for updating, since multiple updates
+    ///      with the same timestamp will get reverted in CumulativeTwap._update()
+    function update() external {
```

Comment:
should we impl IPriceFeedUpdater? I was confused when checking PriceFeedUpdater

Reply:
Yes, we should.
Fixed in https://github.com/perpetual-protocol/perp-oracle-contract/pull/60